lib/msf/core/post/linux/system.rb

```diff
@@ -102,6 +102,135 @@ def get_sysinfo
     return system_data
   end
 
+  #
+  # Returns all listening services along with their ports
+  # @param portsonly Return the listening ports without their associated service
+  # @return [Hash]
+  # 
+  def get_listening_services(portsonly = false)
+    services = {}
+    begin
+      lines = cmd_exec('netstat -tulpn | wc -l')
+      cmd = "netstat -tulpn | tail -n #{lines - 2} | awk '{print $7}'"
+      cmd << " | cut -f1 -d '/'" if portsonly
+      full = cmd_exec(cmd)
+      full.delete!(':') # Only happens when getting services
+
+      if portsonly
+        ports = []
+        full.split("\n").each do |p|
+          ports << p
+        end
+        ports
+      else
+        full.split("\n").each do |s|
+          split = s.split('/')
+          services[:"#{split[0]}"] = split[1]
+        end
+      end
+
+      services
+    rescue
+      raise 'Could not gather listening ports'
+    end
+  end
+
+  #
+  # Gathers all SUID files on the filesystem.
+  # NOTE: This uses the Linux `find` command. It will most likely take a while to get all files.
+  # Consider specifying a more narrow find path.
+  # @param findpath The path on the system to start searching
+  # @return [Array]
+  def get_suid_files(findpath = '/')
+    begin
+      cmd_exec("find #{findpath} -perm -4000 -print").split("\n")
```

Comment:
This could run for a looooong time, rendering the session useless until completion.

However, as there's no timeout specified in this `cmd_exec` call, it's more likely that it will run for a while (default 30 seconds?) then die, returning nothing.

It may be worth adding an optional `-xdev` argument.

Reply:
Maybe using something like `timeout 30s find #{findpath} -perm -4000 -print -xdev` would solve the run time problem? It would probably decrease the reliability, but it could solve the problem on most systems.

Reply:
FWIW, the `timeout(1)` you're thinking of is from GNU coreutils. There are a fair number of (frequently embedded) Linux systems without the entire suite of tools. BusyBox `timeout` (if you even have it) takes `-t` for seconds.